rueidislimiter/limiter.go

```diff
@@ -0,0 +1,208 @@
+package rueidislimiter
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/redis/rueidis"
+)
+
+type Result struct {
+	// Allowed indicates if the request is allowed under the rate limit
+	Allowed bool
+	// Remaining is the number of remaining tokens in the current window
+	Remaining int64
+	// ResetAtMs is the Unix timestamp in milliseconds at which the rate limit will reset
+	ResetAtMs int64
+}
+
+type RateLimiterClient interface {
+	Check(ctx context.Context, identifier string) (Result, error)
+	Allow(ctx context.Context, identifier string) (Result, error)
+	AllowN(ctx context.Context, identifier string, n int64) (Result, error)
+}
+
+const PlaceholderPrefix = "rueidislimiter"
+
+type rateLimiter struct {
+	client rueidis.Client
+	key    string
+	limit  int
+	window time.Duration
+}
+
+type RateLimiterOption struct {
+	// ClientBuilder can be used to modify rueidis.Client used by RateLimiter
+	ClientBuilder func(option rueidis.ClientOption) (rueidis.Client, error)
+	ClientOption  rueidis.ClientOption
+	KeyPrefix     string
+	Limit         int
+	Window        time.Duration
+}
+
+func NewRateLimiter(option RateLimiterOption) (rlc RateLimiterClient, err error) {
+	if option.Window < time.Millisecond {
+		option.Window = time.Millisecond
+	}
+	if option.Limit <= 0 {
+		option.Limit = 1
+	}
+	if option.KeyPrefix == "" {
+		option.KeyPrefix = PlaceholderPrefix
+	}
+	rl := &rateLimiter{
+		limit:  option.Limit,
+		window: option.Window,
+	}
+	if option.ClientBuilder != nil {
+		rl.client, err = option.ClientBuilder(option.ClientOption)
+	} else {
+		rl.client, err = rueidis.NewClient(option.ClientOption)
+	}
+	if err != nil {
+		return nil, err
+	}
+	rl.key = option.KeyPrefix
+	return rl, nil
+}
+
+func (l *rateLimiter) Check(ctx context.Context, identifier string) (Result, error) {
```

Comment:
Is using AllowN(ctx, 0) to replace this a good idea? We can maintain less code by doing that.

Reply:
This is a great idea, not it should be more easy to follow and mantain